### ISSUE TITLE
Default channel bindings cause auth failure

### ISSUE THREAD

**Issue by jborean93** [gssapi/gss-ntlmssp#25]
I've found that when no channel bindings are specified, gss-ntlmssp sets `MSV_AV_CHANNEL_BINDINGS` to be `00000000000000000000000000000000`. This causes issues on SSPI when it is set to process the channel bindings token. Typically most MS services that deal with channel bindings have 3 settings

* Ignore - Do not validate the channel bindings if present
* Accept - Validate the bindings but don't fail if not present
* Required - Require and validate the bindings

I've found that when using gss-ntlmssp against an IIS site that is set to either accept or require channel bindings in it's auth configuration, it will fail when processing the Authentication token.

While this doesn't cover the actual Windows setup, you can use the following script to test authentication against an IIS site that is configured to "accept" channel bindings

```python
import base64
import gssapi
import ssl
import urllib3

from cryptography import x509
from cryptography.hazmat.backends import default_backend
from cryptography.hazmat.primitives import hashes


def get_channel_bindings(server):
    cert_pem = ssl.get_server_certificate((server, 443))
    b_cert = ssl.PEM_cert_to_DER_cert(cert_pem)
    cert = x509.load_der_x509_certificate(b_cert, default_backend())

    hash_algorithm = cert.signature_hash_algorithm

    if hash_algorithm.name in ['md5', 'sha1']:
        digest = hashes.Hash(hashes.SHA256(), default_backend())
    else:
        digest = hashes.Hash(hash_algorithm, default_backend())

    digest.update(b_cert)
    b_cert_hash = digest.finalize()

    b_app_data = b"tls-server-end-point:" + b_cert_hash

    return gssapi.raw.ChannelBindings(application_data=b_app_data)


server = '192.168.56.6'
url = 'https://%s/' % server

# It's a pain to get the underlying socket of a request in Python so we just manually get the cert ourselves
channel_bindings = get_channel_bindings(server)
channel_bindings = None

warnings.simplefilter('ignore', category=urllib3.exceptions.InsecureRequestWarning)
http = urllib3.PoolManager(cert_reqs='CERT_NONE')

ntlm = gssapi.OID.from_int_seq('1.3.6.1.4.1.311.2.2.10')

username = gssapi.Name('vagrant', name_type=gssapi.NameType.user)
cred = gssapi.raw.acquire_cred_with_password(username, b'vagrant', usage='initiate', mechs=[ntlm]).creds
c = gssapi.SecurityContext(name=gssapi.Name('http@test', name_type=gssapi.NameType.hostbased_service),
                           creds=cred, mech=ntlm, usage='initiate', channel_bindings=channel_bindings)

out_token = base64.b64encode(c.step()).decode()
print("NTLM Negotiate: %s" % out_token)

resp = http.request('GET', url, headers={
    'Authorization': 'Negotiate %s' % out_token,
})

in_token = resp.headers['WWW-Authenticate'].split(' ')[1]
print("NTLM Challenge: %s" % in_token)

out_token = base64.b64encode(c.step(base64.b64decode(in_token))).decode()
print("NTLM Authenticate: %s" % out_token)

resp = http.request('GET', url, headers={
    'Authorization': 'Negotiate %s' % out_token,
})

if resp.status == 401:
    raise Exception("Authentication failed")
```

If you were to comment out `channel_bindings = None` then the authentication works just fine but when not setting the bindings the authentication fails. Here is the NTLM authentication token that was sent in this exchange

```yaml
MessageType: NtlmAuthenticate (3)
Signature: "NTLMSSP\0"
Data:
  LmChallengeResponseFields:
    Len: 0
    MaxLen: 0
    BufferOffset: 72
  NtChallengeResponseFields:
    Len: 240
    MaxLen: 240
    BufferOffset: 72
  DomainNameFields:
    Len: 0
    MaxLen: 0
    BufferOffset: 0
  UserNameFields:
    Len: 14
    MaxLen: 14
    BufferOffset: 312
  WorkstationFields:
    Len: 26
    MaxLen: 26
    BufferOffset: 326
  EncryptedRandomSessionKeyFields:
    Len: 16
    MaxLen: 16
    BufferOffset: 352
  NegotiateFlags:
    raw: 3800728085
    flags:
    - NTLMSSP_NEGOTIATE_56 (2147483648)
    - NTLMSSP_NEGOTIATE_KEY_EXCH (1073741824)
    - NTLMSSP_NEGOTIATE_128 (536870912)
    - NTLMSSP_NEGOTIATE_VERSION (33554432)
    - NTLMSSP_NEGOTIATE_TARGET_INFO (8388608)
    - NTLMSSP_NEGOTIATE_EXTENDED_SESSIONSECURITY (524288)
    - NTLMSSP_TARGET_TYPE_SERVER (131072)
    - NTLMSSP_NEGOTIATE_ALWAYS_SIGN (32768)
    - NTLMSSP_NEGOTIATE_NTLM (512)
    - NTLMSSP_NEGOTIATE_SIGN (16)
    - NTLMSSP_REQUEST_TARGET (4)
    - NTLMSSP_NEGOTIATE_UNICODE (1)
  Version:
    Major: 6
    Minor: 2
    Build: 0
    Reserved: '000000'
    NTLMRevision: 15
  MIC:
  Payload:
    LmChallengeResponse:
    NtChallengeResponse:
      ResponseType: NTLMv2
      NTProofStr: 55F53974A37F645DC6EAE24F616DCE99
      ClientChallenge:
        RespType: 1
        HiRespType: 1
        Reserved1: 0
        Reserved2: 0
        TimeStamp: '2020-06-06T04:21:11.7360608Z'
        ChallengeFromClient: F108B844B11ED8BD
        Reserved3: 0
        AvPairs:
        - AvId: MSV_AV_NB_COMPUTER_NAME (1)
          Value: WIN-OA4NAHNUCHT
        - AvId: MSV_AV_NB_DOMAIN_NAME (2)
          Value: WIN-OA4NAHNUCHT
        - AvId: MSV_AV_DNS_COMPUTER_NAME (3)
          Value: WIN-OA4NAHNUCHT
        - AvId: MSV_AV_DNS_DOMAIN_NAME (4)
          Value: WIN-OA4NAHNUCHT
        - AvId: MSV_AV_FLAGS (6)
          Value:
            raw: 0
            flags: []
        - AvId: MSV_AV_TIMESTAMP (7)
          Value: '2020-06-06T04:21:11.7360608Z'
        - AvId: MSV_AV_TARGET_NAME (9)
          Value: test
        - AvId: MSV_AV_CHANNEL_BINDINGS (10)
          Value: '00000000000000000000000000000000'
        - AvId: MSV_AV_EOL (0)
          Value:
        Reserved4: 0
    DomainName:
    UserName: vagrant
    Workstation: JBOREAN-LINUX
    EncryptedRandomSessionKey: F0B7AE8F117C4B596E4DA88D35500A05
  SessionKey: Failed to derive
RawData: 4E544C4D53535000030000000000000048000000F000F0004800000000000000000000000E000E00380100001A001A0046010000100010006001000015828AE2060200000000000F55F53974A37F645DC6EAE24F616DCE990101000000000000E0E5F4E8B93BD601F108B844B11ED8BD0000000001001E00570049004E002D004F00410034004E00410048004E00550043004800540002001E00570049004E002D004F00410034004E00410048004E00550043004800540003001E00570049004E002D004F00410034004E00410048004E00550043004800540004001E00570049004E002D004F00410034004E00410048004E005500430048005400060004000000000007000800E0E5F4E8B93BD6010900080074006500730074000A001000000000000000000000000000000000000000000000000000760061006700720061006E0074004A0042004F005200450041004E002D004C0049004E0055005800F0B7AE8F117C4B596E4DA88D35500A05
```

The [MS-NLMP docs](https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-nlmp/f9e6fbc4-a953-4f24-b229-ccdcc213b9ec) for this behaviour are somewhat vague but it states the checks occur when

> If the AUTHENTICATE_MESSAGE contains a nonzero MsvAvChannelBindings AV_PAIR

I originally understood a nonzero pair meaning the value isn't `00000000000000000000000000000000` but it seems to be when the value isn't set at all.

**Comment by jborean93** [comment]
Also I was trying to figure out why gss-ntlmssp was sending an empty bindings token when using it through SPNEGO with actual bindings but it turns out SPNEGO in MIT krb5 only passed the channel bindings along [only recently](https://github.com/krb5/krb5/commit/d16325a24c34ec9a5f6fb4910987f162e0d4d9cd).

**Comment by simo5** [comment]
Sorry I am confused.
Are you having issues with an all 0 binding when CBs are not required ? Or also when they can be ignored ?

**Comment by simo5** [comment]
I am asking becuase MS-NLMP says:
"An all-zero value of the hash is used to indicate absence of channel bindings"

**Comment by jborean93** [comment]
> Are you having issues with an all 0 binding when CBs are not required ? Or also when they can be ignored ?

If they are completely ignored then the value doesn't matter at all, it can be 0's, 1's, or some other garbage and the acceptor doesn't care.

If they are set to accept (but still not require) then the current gss-ntlmssp behaviour fails because the presence of this AV_PAIR has it validating the actual channel binding token hash against the 0'd value.

If it's set to require, then it will fail as expected.

**Comment by simo5** [comment]
Sigh, the docs clearly say an all 0 should be treated as no CBs :-(

**Comment by jborean93** [comment]
Agreed but it's definitely failing when I test it using gss-ntlmssp with the all 0 value as well as a Python client I have. As soon as I remove the whole AV_PAIR the Python client works just fine against the IIS site.

**Comment by jborean93** [comment]
Because this is never easy, I've found that SSPI acts just like gss-ntlmssp when no channel bindings are sent, i.e. `MsvAvChannelBindings.AvPair.Value == 00000000000000000000000000000000`. It has the same problem when trying to talk to an IIS site that is set to 'Accept' (but not 'Require') channel bindings. This means that maybe something was wrong in my assumption (even though IIS acts fine with no MsvAvChannelBindings AV_PAIR in that mode).

I've opened https://social.msdn.microsoft.com/Forums/en-US/bbbbf50b-9fee-4f91-9eb2-8e24886fc273/msnlmp-empty-channel-bindings-hash?forum=os_windowsprotocols to try and track down what may be happening.
